qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/extensions/renarde/UriNamespaceResolverSupport.java

```diff
@@ -0,0 +1,150 @@
+/*******************************************************************************
+* Copyright (c) 2023 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.qute.jdt.internal.extensions.renarde;
+
+import static com.redhat.qute.jdt.internal.QuteJavaConstants.JAVA_LANG_OBJECT_TYPE;
+import static com.redhat.qute.jdt.internal.extensions.renarde.RenardeJavaConstants.RENARDE_CONTROLLER_TYPE;
+import static com.redhat.qute.jdt.internal.extensions.renarde.RenardeJavaConstants.RENARDE_CONTROLLER_WITH_USER_TYPE;
+
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.Flags;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeHierarchy;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.search.SearchMatch;
+import org.eclipse.jdt.core.search.SearchPattern;
+
+import com.redhat.qute.commons.datamodel.resolvers.ValueResolverInfo;
+import com.redhat.qute.commons.datamodel.resolvers.ValueResolverKind;
+import com.redhat.qute.jdt.template.datamodel.AbstractDataModelProvider;
+import com.redhat.qute.jdt.template.datamodel.SearchContext;
+import com.redhat.qute.jdt.utils.JDTTypeUtils;
+
+/**
+ * uri, uriabs renarde support.
+ * 
+ * @author Angelo ZERR
+ * 
+ * @see https://github.com/quarkiverse/quarkus-renarde/blob/main/docs/modules/ROOT/pages/index.adoc#obtaining-a-uri-in-qute-views
+ *
+ */
+public class UriNamespaceResolverSupport extends AbstractDataModelProvider {
+
+	private static final Logger LOGGER = Logger.getLogger(UriNamespaceResolverSupport.class.getName());
+
+	private static final String URI_NAMESPACE = "uri";
+
+	private static final String URIABS_NAMESPACE = "uriabs";
+
+	@Override
+	public void beginSearch(SearchContext context, IProgressMonitor monitor) {
+		IJavaProject javaProject = context.getJavaProject();
+		IType type = JDTTypeUtils.findType(javaProject, RENARDE_CONTROLLER_TYPE);
+		if (type != null) {
+			try {
+				// Find all classes which extends 'io.quarkiverse.renarde.Controller'
+				collectRenardeController(type, context, monitor);
+				// Find all classes which extends
+				// 'io.quarkiverse.renarde.oidc.ControllerWithUser'
+				collectRenardeController(JDTTypeUtils.findType(javaProject, RENARDE_CONTROLLER_WITH_USER_TYPE), context,
+						monitor);

```

Comment:
Can you play around with the following snippet ? It seems to fix the test for me :


```diff
diff --git a/qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/QuteProjectTest.java b/qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/QuteProjectTest.java
index 06a44eb..3c5a2de 100644
--- a/qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/QuteProjectTest.java
+++ b/qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/QuteProjectTest.java
@@ -120,7 +120,8 @@ public class QuteProjectTest {
 			};
 			IProgressMonitor monitor = new NullProgressMonitor();
 			JavaCore.run(runnable, null, monitor);
-			waitForBackgroundJobs(monitor);			
+			waitForBackgroundJobs(monitor);
+			org.eclipse.jdt.ls.core.internal.JobHelpers.waitUntilIndexesReady();
 		}
 		// Collect Quarkus properties from the "hibernate-orm-resteasy" project. It
 		// should collect Quarkus properties from given JAR:
diff --git a/qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/extensions/renarde/UriNamespaceResolverSupport.java b/qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/extensions/renarde/UriNamespaceResolverSupport.java
index 437b680..907fa92 100644
--- a/qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/extensions/renarde/UriNamespaceResolverSupport.java
+++ b/qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/extensions/renarde/UriNamespaceResolverSupport.java
@@ -58,10 +58,6 @@ public class UriNamespaceResolverSupport extends AbstractDataModelProvider {
 			try {
 				// Find all classes which extends 'io.quarkiverse.renarde.Controller'
 				collectRenardeController(type, context, monitor);
-				// Find all classes which extends
-				// 'io.quarkiverse.renarde.oidc.ControllerWithUser'
-				collectRenardeController(JDTTypeUtils.findType(javaProject, RENARDE_CONTROLLER_WITH_USER_TYPE), context,
-						monitor);
 			} catch (JavaModelException e) {
 				LOGGER.log(Level.SEVERE, "Error while collecting Renarde Controller.", e);
 			}
```

I'm not sure if I put it in the best place, but the idea is once a project has successfully imported, the index could still be building in the background, so before creating any queries, we should make sure the index loaded completely.

Reply:
Thanks so much @rgrunber for your great suggestion. It works like a charm now. No need to use some hack. Thanks!